core/blockchain.go

```diff
@@ -1626,3 +1637,90 @@ func (bc *BlockChain) SubscribeChainSideEvent(ch chan<- ChainSideEvent) event.Su
 func (bc *BlockChain) SubscribeLogsEvent(ch chan<- []*types.Log) event.Subscription {
 	return bc.scope.Track(bc.logsFeed.Subscribe(ch))
 }
+
+// GetShardState retrives sharding state given block hash and block number
+func (bc *BlockChain) GetShardState(hash common.Hash, number uint64) types.ShardState {
+	if cached, ok := bc.shardStateCache.Get(hash); ok {
+		shardState := cached.(types.ShardState)
+		return shardState
+	}
+	shardState := rawdb.ReadShardState(bc.db, hash, number)
+	if shardState == nil {
+		return nil
+	}
+	bc.shardStateCache.Add(hash, shardState)
+	return shardState
+}
+
+// GetShardStateByNumber retrieves sharding state given the block number
+func (bc *BlockChain) GetShardStateByNumber(number uint64) types.ShardState {
+	hash := rawdb.ReadCanonicalHash(bc.db, number)
+	if hash == (common.Hash{}) {
+		return nil
+	}
+	return bc.GetShardState(hash, number)
+}
+
+// GetShardStateByHash retrieves the shard state given the blockhash, return nil if not exist
+func (bc *BlockChain) GetShardStateByHash(hash common.Hash) types.ShardState {
+	number := bc.hc.GetBlockNumber(hash)
+	if number == nil {
+		return nil
+	}
+	return bc.GetShardState(hash, *number)
+}
+
+// GetRandSeedByNumber retrieves the rand seed given the block number, return 0 if not exist
+func (bc *BlockChain) GetRandSeedByNumber(number uint64) int64 {
+	header := bc.GetHeaderByNumber(number)
+	if header == nil {
+		return 0
+	}
+	return int64(header.RandSeed)
+}
+
+// GetNewShardState will calculate (if not exist) and get the new shard state for epoch block or nil if block is not epoch block
+// epoch block is where the new shard state stored
+func (bc *BlockChain) GetNewShardState(block *types.Block) types.ShardState {
+	hash := block.Hash()
+	number := block.NumberU64()
+	// just ignore non-epoch block
+	if !CheckEpochBlock(number) {
+		return nil
+	}
+	shardState := bc.GetShardState(hash, number)
+	if shardState == nil {
+		epoch := GetEpochFromBlockNumber(number)
+		shardState = CalculateNewShardState(bc, epoch)
+		bc.shardStateCache.Add(hash, shardState)
+	}
+	return shardState
+}
+
+// ValidateNewShardState validate whether the new shard state root matches
+func (bc *BlockChain) ValidateNewShardState(block *types.Block) error {
```

Comment:
the return value is not clear to me?  return nil means okay?  please be more explicit.

Reply:
not all block contains shard value, in this case return nil. If the block contains shard value, return error if root hash does not match. only epoch block contains shard state hash.